### PR TITLE
Generate bill run

### PR DESCRIPTION
**Pull request by StuAA78**
We now have everything we need to generate the bill run and populate the bill run level stats. This branch applies the following:

- Debit, credit and zero value invoices are summarised;
- Flags are set accordingly for zero value and deminimis invoices;
- The bill run status is set to `generated` when generation is complete.

### CODE REVIEW

File: test/services/validate_bill_run.service.test.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    payload = GeneralHelper.cloneObject(requestFixtures.simple)
+  })
+
+  afterEach(async () => {
+    Sinon.restore()
+  })
+
+  describe('When a valid bill run ID is supplied', () => {
+    beforeEach(async () => {
+      Sinon.stub(RulesService, 'go').returns(rulesServiceResponse)
```
Comment (Cruikshanks) [inline test/services/validate_bill_run.service.test.js:55]: _I think_ for the bill run to be valid it just needs to have a transaction linked to it.

If that's the case could we drop Sinon and stubbing altogether and just use the `TransactionHelper`?

File: test/services/validate_bill_run.service.test.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    it('returns true', async () => {
+      await CreateTransactionService.go(payload, billRun.id, authorisedSystem, regime)
+
+      const result = await ValidateBillRunService.go(billRun.id)
+
+      expect(result).to.equal(true)
+    })
+  })
+
+  describe('When an invalid bill run ID is supplied', () => {
```
Comment (Cruikshanks) [inline test/services/validate_bill_run.service.test.js:68]: Do we need a test to cover a bill run with no transactions?

File: test/support/helpers/rules_service.helper.js [hunk 3]
```diff
@@ -21,6 +21,23 @@ class RulesServiceHelper {
...
   }
+
+  /**
+   * Mock the rules service to return a specific charge value.
+   *
+   * @param {object} Sinon The instance of Sinon used in test.
+   * @param {object} rulesServiceResponse The rules service response fixture used in test.
+   * @param {integer} chargeValue The charge value to be returned from the rules service.
+   */
+  static mockValue (Sinon, RulesService, rulesServiceResponse, chargeValue) {
```
Comment (Cruikshanks) [inline test/support/helpers/rules_service.helper.js:32]: ❤️ 🏆 👏 This is going to be so handy!

**Review by Cruikshanks — CHANGES_REQUESTED** [review]
There are some things raised we should discuss outside of this PR that will help keep our code and tests consistent.

I've gone off on one about comments (sorry!) but apart from that, it's just some suggestions about the Validate service tests. 😁

File: app/services/generate_bill_run.service.js [hunk 4]
```diff
@@ -32,31 +33,26 @@ class GenerateBillRunService {
...
+      count: invoices.length,
+      value: this._sumInvoices(invoices)
     }
   }
 
-  static async _setGeneratingStatus (billRun) {
-    await billRun.$query()
-      .patch({ status: 'generating' })
+  static _sumInvoices (invoices) {
+  /**
```
[the remark this answers is not in the document]
Comment (StuAA78) [reply app/services/generate_bill_run.service.js:51]: Yes that's fair enough! I started using `/** */` as VS Code automatically adds a new comment line as you go but I can see that it's worth preserving the difference.

File: test/services/generate_bill_run.service.test.js [hunk 5]
```diff
@@ -66,77 +59,148 @@ describe('Generate Bill Run Summary service', () => {
 
   describe('When a valid bill run ID is supplied', () => {
     beforeEach(async () => {
-      Sinon.stub(RulesService, 'go').returns(chargeFixtures.simple.rulesService)
+      rulesServiceStub = Sinon.stub(RulesService, 'go').returns(rulesServiceResponse)
```
[the remark this answers is not in the document]
Comment (StuAA78) [reply test/services/generate_bill_run.service.test.js:62]: I'd be happy with that, yes -- makes sense to use Sinon wherever we can unless we absolutely need to mock responses (or there's something we can do with Nock that we can't with Sinon)

File: app/services/generate_bill_run.service.js [hunk 6]
```diff
@@ -95,20 +91,55 @@ class GenerateBillRunService {
...
   }
 
   static async _summariseZeroValueInvoices (billRun, trx) {
-    return billRun.$relatedQuery('invoices', trx)
+    const zeroCount = await billRun.$relatedQuery('invoices', trx)
       .modify('zeroValue')
-      .patch({ summarised: true })
+      .patch({ zeroValueInvoice: true })
+
+    await billRun.$query(trx)
```
[the remark this answers is not in the document]
Comment (StuAA78) [reply app/services/generate_bill_run.service.js:130]: Good shout flagging this -- it's an error! `billRun.zeroCount` is the number of zero value _transactions_ in the bill run, not invoices, and this is incorrectly overwriting it with the number of zero value invoices. I'll change this and update the test to ensure it's counting transactions, not invoices.

File: test/services/validate_bill_run.service.test.js (hunk 1 above)
Comment (StuAA78) [reply test/services/validate_bill_run.service.test.js:55]: The validation relies on the bill run-level credit/debit/zero value counts, which aren't updated by `TransactionHelper`. So we do still need to use `CreateTransactionService` (unless we change how the bill run identifies itself as empty).

File: test/services/validate_bill_run.service.test.js (hunk 2 above)
Comment (StuAA78) [reply test/services/validate_bill_run.service.test.js:68]: Yes, good shout 👍

File: app/controllers/presroc/bill_runs.controller.js [hunk 7]
```diff
@@ -16,11 +21,15 @@ class BillRunsController {
...
-    await GenerateBillRunService.go(req.params.billRunId)
+    await ValidateBillRunService.go(req.params.billRunId)
+    GenerateBillRunService.go(req.params.billRunId)
 
     return h.response().code(204)
   }
+
+  static async _startBillRunGeneration (billRunId) {
+    await GenerateBillRunService.go(billRunId)
+  }
```
[the remark this answers is not in the document]
Comment (StuAA78) [reply app/controllers/presroc/bill_runs.controller.js:32]: Yes it can! 🤦

**Review by Cruikshanks — APPROVED** [review]
lgtm

**Review by Cruikshanks — APPROVED** [review]

**Pull request merged**